File: ts/ex.2.2.MakeAChain(longText).ts
import { StringTemplate, RegxExtractor } from './utils';
import { LLMsStore } from './LLMsModel';
import { TextFile } from './file';

// Initialize the LLMsStore
const store = new LLMsStore();

// Define the system prompt
const systemPrompt = `I will provide pieces of the text along with prior summarizations.
Your task is to read each new text snippet and add new summarizations accordingly.
You should reply summarizations only, without any additional information.

## Your Reply Format Example
\`\`\`summarization
- This text shows ...
\`\`\``;

// Add vendors and LLM models
const openaiVendor = store.addNewOpenAIVendor(process.env.OPENAI_API_KEY);
const chatGPT4oMini = store.addNewChatGPT4oMini(openaiVendor.get_id());
chatGPT4oMini.system_prompt = systemPrompt;

// const ollamaVendor = store.addNewOllamaVendor();
// const gemma2 = store.addNewGemma2(ollamaVendor.get_id());
// gemma2.system_prompt = systemPrompt;

// const phi3 = store.addNewPhi3(ollamaVendor.get_id());
// phi3.system_prompt = systemPrompt;

// const llama32 = store.addNewLlama(ollamaVendor.get_id());
// llama32.system_prompt = systemPrompt;

// Add a message template
const msgTemplate = store.add_new_obj(
    new StringTemplate(`Please reply summarizations in {}, and should not over {} words.
## Text Snippet
\`\`\`text
{}
\`\`\`
## Previous Summarizations
\`\`\`summarization
{}
\`\`\``)
);

// Add a regex extractor
const resExt = store.add_new_obj(
    new RegxExtractor('```summarization\\n([\\s\\S]*?)\\n```')
);

// Test summarization function
async function* testSummary(
    llm = chatGPT4oMini,
    filePath: string,
    limitWords: number = 1000,
    chunkLines: number = 100,
    overlapLines: number = 30
) {
    let preSummarization: string | null = null;
    const textFile = new TextFile(filePath, chunkLines, overlapLines);

    for (const chunk of textFile) {
        const msg = msgTemplate.call(['Japanese', limitWords, chunk.join('\n'), preSummarization]);
        const output = await llm.acall(msg);
        const result = resExt.call(output);

        preSummarization = result;
        yield result;
    }
}

// Custom chain example
function compose(...funcs: Function[]): Function {
    return (...args: any[]) =>
      funcs.reduce((acc, func) => ( func(acc) ), args);
  }
  
const chainList = [msgTemplate, chatGPT4oMini, resExt];
const chain = compose(
    msgTemplate.call.bind(msgTemplate),
    chatGPT4oMini.acall.bind(chatGPT4oMini),
    resExt.acall.bind(resExt)
);

// Example usage
let preSummarization = '';
const textFile = new TextFile('../The Adventures of Sherlock Holmes.txt', 100, 30);

for (const chunk of textFile) {
    // const summarization = await chain(['Japanese', 100, chunk.join('\n'), preSummarization]);
    // preSummarization = summarization;
    console.log(await resExt.acall(chatGPT4oMini.acall(msgTemplate.call(['Japanese', 100, chunk.join('\n'), preSummarization]))));
    break; // Remove this break to process the entire file
}

// Batch processing for multiple files
// async function processFiles(files: string[]) {
//     for (const file of files) {
//         const summarizer = testSummary(chatGPT4oMini, file);
//         let combinedSummarization = '';
        
//         for await (const summarization of summarizer) {
//             combinedSummarization += summarization;

//             // Save to individual file
//             const outputFile = file.replace('file', 'output');
//             const fs = require('fs');
//             fs.appendFileSync(outputFile, summarization);
//         }

//         // Save combined summarizations
//         const allInOneFile = 'allinone.txt';
//         const fs = require('fs');
//         fs.appendFileSync(allInOneFile, `## ${file}\n${combinedSummarization}`);
//     }
// }

File: ex_2_2_MakeAChain(longText).py
import os
from LLMAbstractModel import LLMsStore
from LLMAbstractModel.LLMsModel import Model4LLMs
from LLMAbstractModel.utils import RegxExtractor, StringTemplate, TextFile

store = LLMsStore()

system_prompt = '''I will provide pieces of the text along with prior summarizations.
Your task is to read each new text snippet and add new summarizations accordingly.
You should reply summarizations only, without any additional information.

## Your Reply Format Example
```summarization
- This text shows ...
```'''

vendor = store.add_new_vendor(Model4LLMs.OpenAIVendor)(api_key="OPENAI_API_KEY")#auto check os.environ
llm = chatgpt = store.add_new_llm(Model4LLMs.ChatGPT41Nano)(vendor_id='auto',system_prompt=system_prompt)

msg_template = store.add_new_function(StringTemplate(para=dict(
     string='''
Please reply summarizations in {lang}, and should not over {limit_words} words.


## Text Snippet
```text
{txt}
```
## Previous Summarizations
```summarization
{summary}

```'''
))).build()

res_ext = store.add_new_function(RegxExtractor(para=dict(regx=r"```summarization\s*(.*)\s*\n```")))

def test_summary(llm = llm,
                 f='The Adventures of Sherlock Holmes.txt',
                 limit_words=1000,chunk_lines=100, overlap_lines=30):
    
    pre_summarization = None
    text_file = TextFile(file_path=f, chunk_lines=chunk_lines, overlap_lines=overlap_lines)
    for i,chunk in enumerate(text_file):        
        
        msg    = msg_template(lang='Japanese', limit_words=limit_words,txt='\n'.join(chunk),summary=pre_summarization)
        output = llm(msg)
        output = res_ext(output)

        pre_summarization = output
        yield output


############# make a custom chain 
def compose(*funcs):
    def composed(*args, **kwargs):
        result = funcs[0](*args, **kwargs)  # start with last (innermost) function
        for f in funcs[1:]:
            result = f(result)
        return result
    return composed

chain_list = [msg_template, chatgpt, res_ext]
chain = compose(*chain_list)
# print(LLMsStore.chain_dumps(chain_list))
# print(store.chain_loads(LLMsStore.chain_dumps(chain_list)))

pre_summarization = ''
for i,chunk_lines in enumerate(TextFile(file_path='The Adventures of Sherlock Holmes.txt',
                                  chunk_lines=100, overlap_lines=30)):
        summarization = chain(lang='Japanese', limit_words=100,txt='\n'.join(chunk_lines),summary=pre_summarization)
        pre_summarization = summarization
        print(summarization)
        break

# for i,p in enumerate(['file1','file2','file3','filen']):
#     ts = test_summary(p)
#     res = ''
#     for s in ts:
#         with open(p.replace('file','output'),'a') as f:
#             f.write(s)
#             res += s
#     with open(f'allinone','a') as f:
#         f.write(f'## {p}\n')
#         f.write(res)